src/rpc/blockchain.cpp

```diff
@@ -1612,6 +1613,303 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+            "  \"total_out\": xxxxx,       (numeric) Total amount in all outputs (excluding coinbase and thus reward [ie subsidy + totalfee]).\n"
+            "  \"minfee\": xxxxx,          (numeric) Minimum fee in the block.\n"
+            "  \"maxfee\": xxxxx,          (numeric) Maximum fee in the block.\n"
+            "  \"medianfee\": xxxxx,       (numeric) Truncated median fee in the block.\n"
+            "  \"avgfee\": xxxxx,          (numeric) Average fee in the block.\n"
+            "  \"minfeerate\": xxxxx,      (numeric) Minimum feerate (in satoshis per virtual byte).\n"
+            "  \"maxfeerate\": xxxxx,      (numeric) Maximum feerate (in satoshis per virtual byte).\n"
+            "  \"medianfeerate\": xxxxx,   (numeric) Truncated median feerate (in satoshis per virtual byte).\n"
+            "  \"avgfeerate\": xxxxx,      (numeric) Average feerate (in satoshis per virtual byte).\n"
+            "  \"mintxsize\": xxxxx,       (numeric) Minimum transaction size.\n"
+            "  \"maxtxsize\": xxxxx,       (numeric) Maximum transaction size.\n"
+            "  \"mediantxsize\": xxxxx,    (numeric) Truncated median transaction size.\n"
+            "  \"avgtxsize\": xxxxx,       (numeric) Average transaction size.\n"
+            "}\n"
+            "\nExamples:\n"
+            + HelpExampleCli("getblockstats", "1000 \"minfeerate,avgfeerate\"")
+            + HelpExampleRpc("getblockstats", "1000 \"maxfeerate,avgfeerate\"")
+        );
+
+    LOCK(cs_main);
+
+    int height = request.params[0].get_int();
+    int current_tip = chainActive.Height();
+    if (height < 0) {
+        height = current_tip + height + 1;
+    }
+    if (height < 0) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Target block height %d is negative", height));
+    }
+    if (height > current_tip) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Target block height %d after current tip %d", height, current_tip));
+    }
+
+    std::set<std::string> stats;
+    if (request.params.size() > 1 && !request.params[1].isNull()) {
+        UniValue stats_univalue = request.params[1].get_array();
+        for (unsigned int i = 0; i < stats_univalue.size(); i++) {
+            const std::string stat = stats_univalue[i].get_str();
+            if (valid_stats.count(stat) == 0) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid requested statistic %s", stat));
+            }
+            stats.insert(stat);
+        }
+    } else {
+        stats = valid_stats;
+    }
+
+    const CBlockIndex* pblockindex = chainActive[height];
+    const CBlock block = GetBlockChecked(pblockindex);
+    std::map<std::string, UniValue> map_stats = GetBlockStatsMap(block, pblockindex, stats);
+
+    UniValue ret(UniValue::VOBJ);
+    for (const std::string stat : stats) {
+        ret.pushKV(stat, map_stats[stat]);
+    }
```

Comment:
Could move the "Invalid requested statistic" check to here, ie:

```
    for (auto stat : stats) {
        if (map_stats.count(stat) == 0) {
            throw JSONRPCError(...);
        }
        ret.pushKV(stat, map_stats[stat]);
   }
```

Then you wouldn't need `valid_stats`. Downside is the error response would only come after the valid stats were corrected.

Reply:
Yeah, it could be a nice code simplification, but we would still need valid_stats for the default (or we would need to hardcode the default). Besides, I don't like the downside.
I'm not sure about this, but it's not a blocker, right?